File: tools/verify-architecture.mjs
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";

const forbiddenPaths = [
  "backend/apps",
  "backend/src/totem_hub",
  "frontend/components",
  "frontend/lib",
  "frontend/types"
];

for (const path of forbiddenPaths) {
  if (existsSync(path)) {
    throw new Error(`Forbidden legacy or monolith path exists: ${path}`);
  }
}

const required = ["apps/web", "apps/api-gateway", "packages/shared-kernel", "packages/service-runtime", "services"];
for (const path of required) {
  if (!existsSync(path)) {
    throw new Error(`Required active architecture path is missing: ${path}`);
  }
}

const serviceRequiredSubpaths = [
  "src/domain/entities.ts",
  "src/domain/events.ts",
  "src/domain/errors.ts",
  "src/application/commands.ts",
  "src/application/queries.ts",
  "src/application/use-cases.ts",
  "src/ports/repositories.ts",
  "src/adapters/http/routes.ts",
  "src/adapters/http/business-routes.ts",
  "src/adapters/prisma/repository.ts",
  "src/main.ts",
  "prisma/schema.prisma"
];

for (const service of readdirSync("services")) {
  const servicePath = join("services", service);
  if (!statSync(servicePath).isDirectory()) continue;
  for (const subpath of serviceRequiredSubpaths) {
    const requiredPath = join(servicePath, subpath);
    if (!existsSync(requiredPath)) {
      throw new Error(`${service} missing ${subpath}`);
    }
  }
}

function walk(dir) {
  const entries = readdirSync(dir);
  return entries.flatMap((entry) => {
    const path = join(dir, entry);
    const stat = statSync(path);
    if (stat.isDirectory() && (entry === "node_modules" || entry === "generated" || entry === "dist" || entry === ".next")) {
      return [];
    }
    return stat.isDirectory() ? walk(path) : [path];
  });
}

const files = [...walk("services"), ...walk("apps"), ...walk("packages")].filter((path) => path.endsWith(".ts") || path.endsWith(".tsx"));
const forbiddenCodePatterns = [" as any", " as never", "@ts-ignore", "eslint-disable"];
for (const file of files) {
  const text = readFileSync(file, "utf8");
  for (const pattern of forbiddenCodePatterns) {
    if (text.includes(pattern)) {
      throw new Error(`Forbidden code pattern ${pattern}: ${file}`);
    }
  }
  if (file.includes("/domain/") && (text.includes("@prisma/client") || text.includes("node:http"))) {
    throw new Error(`Domain layer imports infrastructure: ${file}`);
  }
  if (file.includes("/application/") && text.includes("@prisma/client")) {
    throw new Error(`Application layer imports prisma: ${file}`);
  }
  if (file.includes("/ui/") && text.includes("fetch(")) {
    throw new Error(`UI layer performs direct fetch: ${file}`);
  }
}

console.log("[architecture] ok: ddd hexagonal microservices");

File: tools/verify_architecture.py
from __future__ import annotations

from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]

FORBIDDEN_PATHS = [
    "backend/apps",
    "backend/src/totem_hub",
    "frontend/components",
    "frontend/lib",
    "frontend/types",
]

REQUIRED_PATHS = [
    "apps/web",
    "apps/api-gateway",
    "packages/shared-kernel",
    "packages/service-runtime",
    "services/identity",
    "services/tenancy",
    "services/catalog",
    "services/itineraries",
    "services/trips",
    "services/enrollments",
    "services/payments",
    "services/subscriptions",
    "services/documents",
    "services/rooming",
    "services/notifications",
    "services/assistant",
    "services/support",
    "services/platform",
    "services/audit",
]

SERVICE_REQUIRED_SUBPATHS = [
    "src/domain/entities.ts",
    "src/domain/events.ts",
    "src/domain/errors.ts",
    "src/application/commands.ts",
    "src/application/queries.ts",
    "src/application/use-cases.ts",
    "src/ports/repositories.ts",
    "src/adapters/http/routes.ts",
    "src/adapters/http/business-routes.ts",
    "src/adapters/prisma/repository.ts",
    "src/main.ts",
    "prisma/schema.prisma",
]

FORBIDDEN_CODE_PATTERNS = [
    " as any",
    " as never",
    "@ts-ignore",
    "eslint-disable",
]


def iter_code_files(path: Path):
    for candidate in path.rglob("*"):
        relative_parts = candidate.relative_to(ROOT).parts
        if "node_modules" in relative_parts or "generated" in relative_parts or "dist" in relative_parts or ".next" in relative_parts:
            continue
        if candidate.is_file() and candidate.suffix in {".ts", ".tsx", ".mjs"}:
            yield candidate


def fail(message: str) -> None:
    raise SystemExit(f"[architecture] {message}")


def main() -> None:
    for relative in FORBIDDEN_PATHS:
        if (ROOT / relative).exists():
            fail(f"forbidden path exists: {relative}")

    for relative in REQUIRED_PATHS:
        if not (ROOT / relative).exists():
            fail(f"required path missing: {relative}")

    for service_dir in sorted((ROOT / "services").iterdir()):
        if not service_dir.is_dir():
            continue
        for subpath in SERVICE_REQUIRED_SUBPATHS:
            required = service_dir / subpath
            if not required.exists():
                fail(f"{service_dir.name} missing {subpath}")

    code_roots = [ROOT / "services", ROOT / "apps", ROOT / "packages"]
    for code_root in code_roots:
        for file_path in iter_code_files(code_root):
            text = file_path.read_text(encoding="utf-8")
            normalized = file_path.as_posix()
            for pattern in FORBIDDEN_CODE_PATTERNS:
                if pattern in text:
                    fail(f"forbidden code pattern {pattern!r}: {file_path.relative_to(ROOT)}")
            if "/domain/" in normalized and ("@prisma/client" in text or "node:http" in text):
                fail(f"domain imports infrastructure: {file_path.relative_to(ROOT)}")
            if "/application/" in normalized and "@prisma/client" in text:
                fail(f"application imports prisma: {file_path.relative_to(ROOT)}")
            if "/ui/" in normalized and "fetch(" in text:
                fail(f"ui performs direct fetch: {file_path.relative_to(ROOT)}")

    print("[architecture] ok: ddd hexagonal microservices")


if __name__ == "__main__":
    main()
